### PR TITLE
Add test to validate Pseries HMI error injection with KVM guest active

### PR DESCRIPTION
**Pull request by sathnaga**
Add test to validate Pseries HMI error injection with KVM guest
running in the given cpu core.

Signed-off-by: Satheesh Rajendran <sathnaga@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by sathnaga** [comment]
```
(01/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.normal.tb_residue.power8: CANCEL (23.66 s)
 (02/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.normal.tb_residue.power9: PASS (36.86 s)
 (03/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.normal.hdec_parity_error.power8: CANCEL (26.80 s)
 (04/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.normal.hdec_parity_error.power9: PASS (37.75 s)
 (05/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.with_save.tb_residue.power8: CANCEL (26.77 s)
 (06/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.with_save.tb_residue.power9: PASS (87.48 s)
 (07/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.with_save.hdec_parity_error.power8: CANCEL (24.59 s)
 (08/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.with_save.hdec_parity_error.power9: PASS (87.14 s)
 (09/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.with_suspend.tb_residue.power8: CANCEL (23.73 s)
 (10/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.with_suspend.tb_residue.power9: PASS (57.00 s)
 (11/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.with_suspend.hdec_parity_error.power8: CANCEL (26.70 s)
 (12/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.with_suspend.hdec_parity_error.power9: PASS (57.02 s)
 (13/14) guest_cpu.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-libvirt.ppchmi.positive.power8_compatguest.tb_residue.power9:```

File: libvirt/tests/src/cpu/ppchmi.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    sm = SoftwareManager()
+    if not sm.check_installed("opal-utils") and not sm.install("opal-utils"):
+        test.error("opal-utils package install failed")
+    cpus_list = cpu.cpu_online_list()
+    cpu_idle_state = get_cpu_idle()
+    set_cpu_idle()
+    # Lets use second available host cpu
+    hmi_cpu = cpus_list[1]
+    pir = int(open('/sys/devices/system/cpu/cpu%s/pir' % hmi_cpu).read().strip(), 16)
+    if host_version == 'power9':
```
Comment (balamuruhans) [inline libvirt/tests/src/cpu/ppchmi.py]: we should get the host cpu version the test is running and assert it with host_version from cfg appropriately right?

File: libvirt/tests/src/cpu/ppchmi.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            session.cmd('while true;do echo "no" >/dev/null;done&')
+        if "save" in condition:
+            set_condn("save")
+        if "suspend" in condition:
+            set_condn("suspend")
+
+        # hmi inject
+        logging.debug("Injecting %s HMI on cpu %s", hmi_name, hmi_cpu)
+        logging.debug("HMI Command: %s", hmi_cmd)
+        process.run(hmi_cmd)
```
Comment (balamuruhans) [inline libvirt/tests/src/cpu/ppchmi.py]: dont we have to check for return ? and let us log the injection by setting verbose

File: libvirt/tests/src/cpu/ppchmi.py (hunk 1 above)
Comment (sathnaga) [reply libvirt/tests/src/cpu/ppchmi.py]: we do that here
```

+    if host_version not in cpu.get_cpu_arch():
+        test.cancel("Unsupported Host cpu version")
```

File: libvirt/tests/src/cpu/ppchmi.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    inject_code = params.get("inject_code", "")
+    scom_base = params.get("scom_base", "")
+    hmi_name = params.get("hmi_name", "")
+    hmi_iterations = int(params.get("hmi_iterations", 1))
+
+    if host_version not in cpu.get_cpu_arch():
+        test.cancel("Unsupported Host cpu version")
+    sm = SoftwareManager()
+    if not sm.check_installed("opal-utils") and not sm.install("opal-utils"):
+        test.error("opal-utils package install failed")
```
[the remark this answers is not in the document]
Comment (sathnaga) [reply libvirt/tests/src/cpu/ppchmi.py]: Sure...

File: libvirt/tests/src/cpu/ppchmi.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def set_condn(condn, recover=False):
+        """
+        Set/reset guest state/action
+        :param condn: Guest state change/action
```
[the remark this answers is not in the document]
Comment (sathnaga) [reply libvirt/tests/src/cpu/ppchmi.py]: Sure...

File: libvirt/tests/cfg/cpu/ppchmi.cfg [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                - hdec_parity_error:
+                    inject_code = 0002080000000000
+                    hmi_name = "TFMR HDEC parity error"
+                    variants:
+                        - power8:
+                            host_version = 'power8'
+                            scom_base = '0x10013281'
+                        - power9:
+                            host_version = 'power9'
+                            scom_base = '0x20010A84'
```
[the remark this answers is not in the document]
Comment (sathnaga) [reply libvirt/tests/cfg/cpu/ppchmi.cfg]: i will give it a try....

File: libvirt/tests/cfg/cpu/ppchmi.cfg [hunk 6]
```diff
@@ -0,0 +1,44 @@
+- ppchmi:
```
[the remark this answers is not in the document]
Comment (sathnaga) [reply libvirt/tests/cfg/cpu/ppchmi.cfg]: Sure..

File: libvirt/tests/src/cpu/ppchmi.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        :param recover: whether to recover given state default: False
+        """
+        if not recover:
+            if condn == "pin_vcpu":
+                for i in range(cur_vcpu):
+                    virsh.vcpupin(vm_name, i, hmi_cpu, "--live",
+                                  ignore_status=False, debug=True)
+                    virsh.emulatorpin(vm_name,  hmi_cpu, "live",
```
[the remark this answers is not in the document]
Comment (sathnaga) [reply libvirt/tests/src/cpu/ppchmi.py]: Sure...

File: libvirt/tests/src/cpu/ppchmi.py (hunk 2 above)
Comment (sathnaga) [reply libvirt/tests/src/cpu/ppchmi.py]: I will check if relevant...

File: libvirt/tests/cfg/cpu/ppchmi.cfg [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                            scom_base = '0x10013281'
+                        - power9:
+                            host_version = 'power9'
+                            scom_base = '0x20010A84'
+            variants:
+                - normal:
+                - with_save:
+                    condn += ,save
+                - with_suspend:
+                    condn += ,suspend
```
Comment (ZongqingQi) [inline libvirt/tests/cfg/cpu/ppchmi.cfg]: I do not know why this cfg cannot be parsed. Tried bootstrap and do not work.

[root@localhost tp-libvirt]# avocado list --vt-type libvirt ppchmi
Unable to resolve reference(s) 'ppchmi' with plugins(s) 'file', 'vt', 'external', try running 'avocado list -V ppchmi' to see the details.

**Comment by sathnaga** [comment]
Test Pending....

**Comment by sathnaga** [comment]
```
(01/16) guest_import.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-qemu.unattended_install.import.import.default_install.aio_native: PASS (47.57 s)
(02/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.normal.power8.tb_residue: CANCEL (27.89 s)
(03/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.normal.power8.hdec_parity_error: CANCEL (5.81 s)
(04/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.normal.power9.tb_residue: PASS (55.28 s)
(05/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.normal.power9.hdec_parity_error: PASS (57.25 s)
(06/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.with_save.power8.tb_residue: CANCEL (28.59 s)
(07/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.with_save.power8.hdec_parity_error: CANCEL (5.85 s)
(08/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.with_save.power9.tb_residue: PASS (111.91 s)
(09/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.with_save.power9.hdec_parity_error: PASS (126.72 s)
(10/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.with_suspend.power8.tb_residue: CANCEL (29.14 s)
(11/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.with_suspend.power8.hdec_parity_error: CANCEL (5.74 s)
(12/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.with_suspend.power9.tb_residue: PASS (67.90 s)
(13/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.with_suspend.power9.hdec_parity_error: PASS (77.55 s)
(14/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.power8_compatguest.power9.tb_residue: PASS (64.12 s)
(15/16) guest_hmi.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.powerpc_hmi.positive.power8_compatguest.power9.hdec_parity_error: PASS (64.21 s)
(16/16) guest_remove.qemu.qcow2.virtio_scsi.smp2.virtio_net.Guest.RHEL.7.devel.ppc64le.powerkvm-libvirt.remove_guest.without_disk: PASS (2.17 s)
```

**Comment by sathnaga** [comment]
@balamuruhans Addressed the review comments

File: libvirt/tests/src/cpu/powerpc_hmi.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        Set/reset guest state/action
+        :param action: Guest state change/action
+        :param recover: whether to recover given state default: False
+        """
+        if not recover:
+            if action == "pin_vcpu":
+                for i in range(cur_vcpu):
+                    virsh.vcpupin(vm_name, i, hmi_cpu, "--live",
+                                  ignore_status=False, debug=True)
+                    virsh.emulatorpin(vm_name,  hmi_cpu, "live",
```
Comment (balamuruhans) [inline libvirt/tests/src/cpu/powerpc_hmi.py]: can we make use of PR https://github.com/avocado-framework/avocado-vt/pull/1550 here

**Comment by balamuruhans** [comment]
@ZongqingQi are you still facing the issue, It is working for me,

File: libvirt/tests/src/cpu/powerpc_hmi.py (hunk 9 above)
Comment (sathnaga) [reply libvirt/tests/src/cpu/powerpc_hmi.py]: @balamuruhans here want to do it confined testing for one single host cpu, whereas stress_events is targeted more for running random pinning events...,it wont match by needs...

**Comment by balamuruhans** [comment]
Thanks @ZongqingQi and @sathnaga, let us merge it.

**Pull request merged**